drivers/arm/ws2812.c

```diff
@@ -0,0 +1,125 @@
+  static THD_FUNCTION(ledsThread, arg) {
+    (void) arg;
+    while(1){
+      spiSend(&WS2812_SPI, PREAMBLE_SIZE + DATA_SIZE + RESET_SIZE, txbuf);
+    }
+  }
+
+  /*
+  * Function used to initialize the driver.
+  *
+  * Starts by shutting off all the LEDs.
+  * Then gets access on the LED_SPI driver.
+  * May eventually launch an animation on the LEDs (e.g. a thread setting the
+  * txbuff values)
+  */
+  void leds_init(void){
+    static SPIConfig spicfg;
+    populateSPIConfig(&spicfg);
+
+    /* MOSI pin*/
+    palSetPadMode(RGB_PORT, RGB_PAD, PAL_MODE_STM32_ALTERNATE_PUSHPULL);
+    for(int i = 0; i < RESET_SIZE; i++)
+      txbuf[DATA_SIZE+i] = 0x00;
+    for (int i=0; i<PREAMBLE_SIZE; i++)
+      txbuf[i] = 0x00;
+    spiAcquireBus(&WS2812_SPI);              /* Acquire ownership of the bus.    */
+    spiStart(&WS2812_SPI, &spicfg);          /* Setup transfer parameters.       */
+    spiSelect(&WS2812_SPI);                  /* Slave Select assertion.          */
+    chThdCreateStatic(LEDS_THREAD_WA, sizeof(LEDS_THREAD_WA), NORMALPRIO, ledsThread, NULL);
+
+    printf("LED Thread going\n");
+
+  }
+  /*
+  * As the trick here is to use the SPI to send a huge pattern of 0 and 1 to
+  * the ws2812b protocol, we use this helper function to translate bytes into
+  * 0s and 1s for the LED (with the appropriate timing).
+  */
+  static uint8_t get_protocol_eq(uint8_t data, int pos){
+    uint8_t eq = 0;
+    if (data & (1 << (2*(3-pos))))
+      eq = 0b1110;
+    else
+      eq = 0b1000;
+    if (data & (2 << (2*(3-pos))))
+      eq += 0b11100000;
+    else
+      eq += 0b10000000;
+    return eq;
+  }
+  void WS2812_init(void) {
+    leds_init();
+  }
+  void ws2812_setleds(LED_TYPE *ledarray, uint16_t number_of_leds) {
+    uint8_t i = 0;
+    while (i < number_of_leds) {
+      set_led_color_rgb(ledarray[i], i);
+      i++;
+    }
+  }
+  /*
+  * If you want to set a LED's color in the RGB color space, simply call this
+  * function with a hsv_color containing the desired color and the index of the
+  * led on the LED strip (starting from 0, the first one being the closest the
+  * first plugged to the board)
+  *
+  * Only set the color of the LEDs through the functions given by this API
+  * (unless you really know what you are doing)
+  */
+  void set_led_color_rgb(LED_TYPE color, int pos){
+    for(int j = 0; j < 4; j++)
+      txbuf[PREAMBLE_SIZE + BYTES_FOR_LED*pos + j] = get_protocol_eq(color.g, j);
+    for(int j = 0; j < 4; j++)
+      txbuf[PREAMBLE_SIZE + BYTES_FOR_LED*pos + BYTES_FOR_LED_BYTE+j] = get_protocol_eq(color.r, j);
+    for(int j = 0; j < 4; j++)
+      txbuf[PREAMBLE_SIZE + BYTES_FOR_LED*pos + BYTES_FOR_LED_BYTE*2+j] = get_protocol_eq(color.b, j);
+  }
+  void set_leds_color_rgb(LED_TYPE color){
+    for(int i = 0; i < NB_LEDS; i++)
+      set_led_color_rgb(color, i);
+  }
+  void ws2812_setleds_rgbw(LED_TYPE *ledarray, uint16_t number_of_leds) {
+  }
+#else
+#endif
```

Comment:
Jack's code to use the timer implementation can go between the else and endif here